yktrace/src/swt.rs

```diff
@@ -55,21 +60,139 @@ struct SWTThreadTracer;
 impl ThreadTracerImpl for SWTThreadTracer {
     #[trace_tail]
     fn stop_tracing(&mut self) -> Result<Box<dyn SirTrace>, InvalidTraceError> {
-        match swt::stop_tracing() {
-            None => Err(InvalidTraceError::InternalError),
-            Some((buf, len)) => Ok(Box::new(SWTSirTrace::from_buf(buf, len)) as Box<dyn SirTrace>)
+        let mut len = 0;
+        let buf = unsafe { yk_swt_stop_tracing_impl(&mut len) };
+        if buf as usize == 0 {
+            Err(InvalidTraceError::InternalError)
+        } else {
+            Ok(Box::new(SWTSirTrace::from_buf(buf, len)) as Box<dyn SirTrace>)
         }
     }
 }
 
 #[trace_head]
 pub fn start_tracing() -> ThreadTracer {
-    swt::start_tracing();
+    unsafe { yk_swt_start_tracing_impl(); }
     ThreadTracer {
         t_impl: Box::new(SWTThreadTracer {})
     }
 }
 
+// Rust translation of the C code removed in https://github.com/softdevteam/ykrustc/pull/121
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct SwtLoc {
+    symbol_name: *const i8,
+    bb_idx: u32,
+}
+
+const TL_TRACE_INIT_CAP: usize = 1024;
+const TL_TRACE_REALLOC_CAP: usize = 1024;
+
+/// The trace buffer.
+#[thread_local]
+static mut TRACE_BUF: *mut SwtLoc = 0 as *mut SwtLoc;
+
+/// The number of elements in the trace buffer.
+#[thread_local]
+static mut TRACE_BUF_LEN: usize = 0;
+
+/// The allocation capacity of the trace buffer (in elements).
+#[thread_local]
+static mut TRACE_BUF_CAP: usize = 0;
+
+/// true = we are tracing, false = we are not tracing or an error occurred.
+#[thread_local]
+static mut TRACING: bool = false;
+
+// =======================
+// The following code may not call any non `#[do_not_trace]` function to prevent an infinite
+// recursion of `__yk_swt_rec_loc` calling a function calling `__yk_swt_rec_loc`.
+
+/// Start tracing on the current thread.
+/// A new trace buffer is allocated and MIR locations will be written into it on
+/// subsequent calls to `yk_swt_rec_loc`. If the current thread is already
+/// tracing, calling this will lead to undefined behaviour.
+#[do_not_trace]
+unsafe fn yk_swt_start_tracing_impl() {
+    TRACE_BUF = libc::calloc(TL_TRACE_INIT_CAP, std::mem::size_of::<SwtLoc>()) as *mut SwtLoc;
+    if TRACE_BUF as usize == 0 {
+        std::intrinsics::abort();
+    }
+
+    TRACE_BUF_CAP = TL_TRACE_INIT_CAP;
+    TRACING = true;
+}
+
+/// Record a location into the trace buffer if tracing is enabled on the current thread.
+#[do_not_trace]
+#[no_mangle]
```

Comment:
Remind me why this is `no_mangle`

Reply:
So that cg_clif can find this function by calling `__yk_swt_rec_loc`.

Reply:
OK.